services/worker-manager/src/providers/azure.js

```diff
@@ -387,19 +383,229 @@ class AzureProvider extends Provider {
+            // chances are our instance has been deleted off band
+            await this.removeWorker({worker});
+          }
+        }
+      }
+    }
+    // failed to lookup resource by name
+    if (!typeData.id) {
+      // we need to create the resource
+      let resourceRequest = await this._enqueue('query', () => client.beginCreateOrUpdate(
+        worker.providerData.resourceGroupName,
+        typeData.name,
+        resourceConfig,
+      ));
+      // track operation
+      await worker.modify(w => {
+        w.providerData[resourceType].operation = resourceRequest.getPollState().azureAsyncOperationHeaderValue;
+      });
+    }
+  }
+
+  /**
+   * provisionResources wraps the process of provisioning worker resources
+   *
+   * This function is expected to be called several times per worker as
+   * resources are created.
+   */
+  async provisionResources({worker}) {
+    try {
+      // IP
+      let ipConfig = {
+        location: worker.providerData.location,
+        publicIPAllocationMethod: 'Dynamic',
+      };
+      await this.provisionResource({
+        worker,
+        client: this.networkClient.publicIPAddresses,
+        resourceType: 'ip',
+        resourceConfig: ipConfig,
+        modifyFn: () => {},
+      });
+      if (!worker.providerData.ip.id) {
+        return worker.state;
+      }
+
+      // NIC
+      let nicConfig = {
+        location: worker.providerData.location,
+        ipConfigurations: [
+          {
+            name: worker.providerData.nic.name,
+            privateIPAllocationMethod: 'Dynamic',
+            subnet: {
+              id: worker.providerData.subnet.id,
+            },
+            publicIPAddress: {
+              id: worker.providerData.ip.id,
+            },
+          },
+        ],
+      };
+      // set up the VM network interface config
+      let nicModifyFunc = (w, nic) => {
+        w.providerData.vm.config.networkProfile.networkInterfaces = [
+          {
+            id: nic.id,
+            primary: true,
+          },
+        ];
+      };
+      await this.provisionResource({
+        worker,
+        client: this.networkClient.networkInterfaces,
+        resourceType: 'nic',
+        resourceConfig: nicConfig,
+        modifyFn: nicModifyFunc,
+      });
+      if (!worker.providerData.nic.id) {
+        return worker.state;
+      }
+
+      // VM
+      // NB: if we ever need disk id for anything, we could get it here
+      await this.provisionResource({
+        worker,
+        client: this.computeClient.virtualMachines,
+        resourceType: 'vm',
+        resourceConfig: workerConfigWithSecrets(worker.providerData.vm.config),
+        modifyFn: () => {},
+      });
+      if (!worker.providerData.vm.id) {
+        return worker.state;
+      }
+      return this.Worker.states.RUNNING;
+    } catch (err) {
+      // we create multiple resources in order to provision a VM
+      // if we catch an error we want to deprovision those resources
```

Comment:
..but this block only reports the error, and doesn't de-provision?

Reply:
Fixed!